src/util/fd_mutex.h

```diff
@@ -0,0 +1,49 @@
+#ifndef HEADER_fd_util_mutex_h
+#define HEADER_fd_util_mutex_h
+
+#include "fd_util_base.h"
+
+/* A simple mutex */
+
+struct fd_mutex {
+  uchar lock;
+};
+typedef struct fd_mutex fd_mutex_t;
+
+FD_PROTOTYPES_BEGIN
+
+static inline void
+fd_mutex_lock( fd_mutex_t * mutex ) {
+#if FD_HAS_THREADS
+  for(;;) {
+    uchar value = mutex->lock;
+    if( FD_LIKELY( !value ) ) {
+      if( FD_LIKELY( FD_ATOMIC_CAS( &mutex->lock, 0, 1 )==0 ) ) return;
+    }
+    FD_SPIN_PAUSE();
+  }
+#else
+  mutex->lock = 1;
+#endif
+}
+
+static inline void
+fd_mutex_unlock( fd_mutex_t * mutex ) {
+  FD_VOLATILE( mutex->lock ) = 0;
```

Comment:
```suggestion
  FD_COMPILER_MFENCE();
  FD_VOLATILE( mutex->lock ) = 0;
```

We need a memory fence before the write to 0 here, so that other threads see the update to the critical section after they see that the lock is free. On x86 architectures this is fine as they have strong ordering guarantees but we should probably put this memory fence in to be safe. See `fd_rwlock.h` as an example.

Reply:
Done